internal/config/variables.go

```diff
@@ -0,0 +1,420 @@
+func (variables *Variables) decodeVariableBlock(block *hcl.Block) hcl.Diagnostics {
+	if (*variables) == nil {
+		(*variables) = Variables{}
+	}
+
+	// TODO krantzinator: A lot of these validations happen twice before now --
+	// when we validate the config on init and on the runner; we probably
+	// don't need it here, too
+	if _, found := (*variables)[block.Labels[0]]; found {
+		return []*hcl.Diagnostic{{
+			Severity: hcl.DiagError,
+			Summary:  "Duplicate variable",
+			Detail:   "Duplicate " + block.Labels[0] + " variable definition found.",
+			Context:  block.DefRange.Ptr(),
+		}}
+	}
+
+	name := block.Labels[0]
+
+	// TODO krantzinator this should happen earlier than the runner
+	// Could be done before any ops are run and fail immediately
+	content, diags := block.Body.Content(variableBlockSchema)
+	if !hclsyntax.ValidIdentifier(name) {
+		diags = append(diags, &hcl.Diagnostic{
+			Severity: hcl.DiagError,
+			Summary:  "Invalid variable name",
+			Detail:   badIdentifierDetail,
+			Subject:  &block.LabelRanges[0],
+		})
+	}
+
+	v := &Variable{
+		Name:  name,
+		Range: block.DefRange,
+	}
+
+	if attr, exists := content.Attributes["description"]; exists {
+		valDiags := gohcl.DecodeExpression(attr.Expr, nil, &v.Description)
+		diags = append(diags, valDiags...)
+	}
+
+	if t, ok := content.Attributes["type"]; ok {
+		tp, moreDiags := typeexpr.Type(t.Expr)
+		diags = append(diags, moreDiags...)
+		if moreDiags.HasErrors() {
+			return diags
+		}
+
+		v.Type = tp
+	}
+
+	if def, ok := content.Attributes["default"]; ok {
+		defaultValue, moreDiags := def.Expr.Value(nil)
+		diags = append(diags, moreDiags...)
+		if moreDiags.HasErrors() {
+			return diags
+		}
+
+		if v.Type != cty.NilType {
+			var err error
+			defaultValue, err = convert.Convert(defaultValue, v.Type)
+			if err != nil {
+				diags = append(diags, &hcl.Diagnostic{
+					Severity: hcl.DiagError,
+					Summary:  "Invalid default value for variable",
+					Detail:   fmt.Sprintf("This default value is not compatible with the variable's type constraint: %s.", err),
+					Subject:  def.Expr.Range().Ptr(),
+				})
+				defaultValue = cty.DynamicVal
+			}
+		}
+
+		v.Values = append(v.Values, VariableAssignment{
+			Source: "default",
+			Value:  defaultValue,
+		})
+
+		// It's possible no type attribute was assigned so lets make sure we
+		// have a valid type otherwise there could be issues parsing the value.
+		if v.Type == cty.NilType {
+			v.Type = defaultValue.Type()
+		}
+	}
+
+	(*variables)[name] = v
+
+	return diags
+}
+
+// Prefix your environment variables with VarEnvPrefix so that Waypoint can see
+// them.
+const VarEnvPrefix = "WP_VAR_"
+
+// Collect values from flags, local var files, and env vars
+func CollectInputVars(vars map[string]string, files []string) ([]*pb.Variable, hcl.Diagnostics) {
+	var diags hcl.Diagnostics
+	ret := []*pb.Variable{}
+
+	// First we'll deal with environment variables, since they have the lowest
+	// precedence.
```

Comment:
Does precedence matter? Since we're retaining the source, do we calculate precedence later? Or does the order of `ret` actually matter?

Reply:
I went back and forth on this a bit. The order of `ret` does matter, currently. The intention with `source` was to use it primarily for error messaging, though it could also be used to build out a specific precedence sort. I'm not tied to either solution, but having things evaluated in precedence-order is what I've got at the moment.